### ISSUE TITLE
[question] package development workflow with custom setting

### ISSUE THREAD

**Issue by martinkunkel2** [conan-io/conan#7199]
We have a package with a hardware dependency (size of available memory). We would like to build the package for different hardware using the package development workflow (https://docs.conan.io/en/latest/developing_packages/package_dev_flow.html). The build step is hardware-dependent.

To achieve this, we introduce a setting for the hardware, in `settings.yml`:
```
hardware: [r-2gb, r-4gb]
```
and a profile `myprofile`, where we specify compiler, arch, etc. and:
```
[settings]
.......
hardware=r-2gb
```
In the `conanfile.py` of our package we write
```
settings = "hardware"
```

Assuming the conanfile.py is in the local directory calling `conan create . --profile myprofile` is successful.

With the package development workflow and first step `conan source` we get
```
conan source . --source-folder tmp/source
ERROR: 'settings.hardware' value not defined
```

Is `conan source` not supported for custom settings? No profile can be specified as argument. The source is independent of the setting.

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
>Is conan source not supported for custom settings? No profile can be specified as argument. The source is independent of the setting.

Yes, the ``source()`` method must be completely agnostic of settings and options, because it is an optimization. The source retrieved by the ``source()`` method will generate a single common "source" folder, containing the source, so such source code should be common to all configurations.

If you need any specific code for different configurations or hardware, you should probably retrieve it in the ``build()`` method. Or you can retrieve both source code variants in the ``source()`` method, and have the ``build()`` method to select the right one to build.

Does this clarify the question? please let me know.

**Comment by martinkunkel2** [comment]
Hi James. Thank you for the answer. 

Our `source()` method is indeed completely agnostic of settings and options. 

The problem is that calling `conan source` still results in an error
```
conan source . --source-folder tmp/source
ERROR: 'settings.hardware' value not defined
```

Here is a minimal example:
```
from conans import ConanFile, tools

class HardwareSettingTestConan(ConanFile):
    name = "HardwareSettingTest"
    version = "1.0.0"

    settings = {"hardware"}

    def source(self):
       print("no setting used in source() method")
```

used conan version 1.26.0

**Comment by tjamsa** [comment]
Hi, I have this same issue. I want to use custom setting defined in `settings.yml`. Setting must have value, so the value can't be `None`. Value for custom setting is set in project specific profile and it works with `conan create` command.

But just like in Martin's example above, `conan source` gives error, because it can't find value for custom setting. There is no way to give profile as argument for the command.

What I have found is that `conan source` can take setting values from `profiles/default`. If I set value for custom setting in  `profiles/default` then `conan source` command works fine. But I do not want to set default value for custom setting, because correct value is set in project specific profile. If `conan source` can already can read the default profile, is it possible to add `--profile` argument for `conan source` command?

**Comment by tjamsa** [comment]
Found possible solution for package development flow with custom setting.

At first run `conan install . --install-folder=tmp/build --profile myProjectProfile`
Then run `conan source . --source-folder=tmp/source --install-folder=tmp/build`
This way `conan source` command finds value for the custom setting from install/build folder.

If this is the correct way to solve this, maybe note could be added to package development flow documentation: https://docs.conan.io/en/latest/developing_packages/package_dev_flow.html ? Idea of running `conan install` before `conan source` was found from here: https://docs.conan.io/en/latest/reference/commands/development/source.html (examples section)

**Comment by memsharded** [comment]
Hi!

Thanks @tjamsa for the solution. Let me explain a little bit the current state:

- ``conan source`` does indeed get the settings at the moment, but it shouldn't use them. This is legacy, we couldn't remove them without breaking users, so we just advise against using settings in the ``source()`` method.
- So the ``source()`` evaluation will try first to fill the settings, and complain if some setting value is not defined.
- The ``conan install`` is also the first command to run before any other local command. It will capture settings and other stuff that is necessary for other local development flow commands as well, like the ``conan source``, ``conan build`` and ``conan package``. So when you are running ``conan source``, the ``hardware`` setting will have the value you used at ``install`` time.

At the moment this would be the recommended flow. I am moving this issue to the docs repo, so it can be improved there to clarify, and we will change the behavior in Conan 2.0.

Thanks for the feedback!
